File: ios/Sources/ACRendering/Views/ActionSetView.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import SwiftUI
import ACCore
import ACActions
import ACAccessibility
import ACFluentUI

struct ActionSetView: View {
    let actions: [CardAction]
    let hostConfig: HostConfig
    var depth: Int = 0

    @Environment(\.actionHandler) var actionHandler
    @Environment(\.actionDelegate) var actionDelegate
    @Environment(\.layoutDirection) var layoutDirection
    @EnvironmentObject var viewModel: CardViewModel

    var body: some View {
        VStack(spacing: CGFloat(hostConfig.actions.buttonSpacing)) {
            Group {
                if orientation == .horizontal {
                    ActionFlowLayout(spacing: CGFloat(hostConfig.actions.buttonSpacing), isRTL: layoutDirection == .rightToLeft) {
                        actionContent
                    }
                } else {
                    VStack(spacing: CGFloat(hostConfig.actions.buttonSpacing)) {
                        actionContent
                    }
                }
            }
            .frame(maxWidth: .infinity, alignment: alignment)

            // Render expanded ShowCard sub-cards inline
            ForEach(showCardActions, id: \.id) { action in
                if case .showCard(let showCardAction) = action,
                   viewModel.isShowCardExpanded(actionId: action.id) {
                    VStack(alignment: .leading, spacing: CGFloat(hostConfig.spacing.default)) {
                        ForEach(Array((showCardAction.card.body ?? []).enumerated()), id: \.offset) { _, element in
                            ElementView(element: element, hostConfig: hostConfig, depth: depth)
                        }
                        // Render sub-card actions if present
                        if let subActions = showCardAction.card.actions, !subActions.isEmpty {
                            ActionSetView(actions: subActions, hostConfig: hostConfig, depth: depth)
                        }
                    }
                    .padding(CGFloat(hostConfig.spacing.padding))
                    .background(
                        RoundedRectangle(cornerRadius: CGFloat(hostConfig.cornerRadius["container"] ?? 4))
                            .fill(Color(hex: hostConfig.containerStyles.emphasis.backgroundColor))
                    )
                    .transition(.opacity.combined(with: .move(edge: .top)))
                }
            }
        }
        .accessibilityContainer(label: "Actions")
    }

    /// All ShowCard actions from the action list
    private var showCardActions: [CardAction] {
        actions.filter {
            if case .showCard = $0 { return true }
            return false
        }
    }

    // MARK: - Overflow logic

    /// Primary actions that fit within maxActions limit
    private var visibleActions: [CardAction] {
        let primary = actions.filter { $0.mode != .secondary }
        let maxActions = hostConfig.actions.maxActions
        if primary.count > maxActions {
            return Array(primary.prefix(maxActions))
        }
        return primary
    }

    /// Overflow actions: explicit secondary + primary actions exceeding maxActions
    private var overflowActions: [CardAction] {
        let primary = actions.filter { $0.mode != .secondary }
        let secondary = actions.filter { $0.mode == .secondary }
        let maxActions = hostConfig.actions.maxActions
        var overflow = [CardAction]()
        if primary.count > maxActions {
            overflow.append(contentsOf: primary.dropFirst(maxActions))
        }
        overflow.append(contentsOf: secondary)
        return overflow
    }

    @ViewBuilder
    private var actionContent: some View {
        ForEach(visibleActions) { action in
            actionButtonView(for: action)
                .fixedSize(horizontal: !isStretch, vertical: false)
                .if(isStretch) { view in
                    view.frame(maxWidth: .infinity)
                }
        }

        if !overflowActions.isEmpty {
            overflowMenu
        }
    }

    @ViewBuilder
    private func actionButtonView(for action: CardAction) -> some View {
        if case .popover(let popoverAction) = action {
            let actionId = popoverAction.id ?? "popover_\(popoverAction.title ?? action.id)"
            ActionButton(action: action, hostConfig: hostConfig) {
                actionHandler.handle(action, delegate: actionDelegate, viewModel: viewModel)
            }
            .sheet(isPresented: viewModel.popoverBinding(actionId: actionId)) {
                PopoverContentView(
                    content: popoverAction.content,
                    title: popoverAction.title,
                    hostConfig: hostConfig,
                    depth: depth
                )
                .environmentObject(viewModel)
                .environment(\.actionHandler, actionHandler)
                .environment(\.actionDelegate, actionDelegate)
            }
        } else {
            ActionButton(action: action, hostConfig: hostConfig) {
                actionHandler.handle(action, delegate: actionDelegate, viewModel: viewModel)
            }
        }
    }

    /// Overflow "..." button that opens a dropdown with secondary actions
    private var overflowMenu: some View {
        Menu {
            ForEach(overflowActions) { action in
                Button(action.title ?? "") {
                    actionHandler.handle(action, delegate: actionDelegate, viewModel: viewModel)
                }
            }
        } label: {
            Text("\u{2026}")
                .padding(.horizontal, CGFloat(hostConfig.spacing.medium))
                .padding(.vertical, CGFloat(hostConfig.spacing.small) * 0.75)
                .foregroundColor(Color(hex: hostConfig.containerStyles.default.foregroundColors.default.default))
                .overlay(
                    RoundedRectangle(cornerRadius: CGFloat(hostConfig.cornerRadius["container"] ?? 4))
                        .stroke(Color(hex: hostConfig.containerStyles.default.foregroundColors.default.subtle), lineWidth: CGFloat(hostConfig.separator.lineThickness))
                )
                .clipShape(RoundedRectangle(cornerRadius: CGFloat(hostConfig.cornerRadius["container"] ?? 4)))
        }
        .accessibilityLabel("More actions")
    }

    // MARK: - Layout helpers

    private var orientation: Orientation {
        hostConfig.actions.actionsOrientation.lowercased() == "vertical" ? .vertical : .horizontal
    }

    private var isStretch: Bool {
        hostConfig.actions.actionAlignment.lowercased() == "stretch"
    }

    private var alignment: Alignment {
        let alignmentStr = hostConfig.actions.actionAlignment.lowercased()
        switch alignmentStr {
        case "center":
            return .center
        case "right":
            return .trailing
        case "stretch":
            return .leading
        default:
            return .leading
        }
    }

    enum Orientation {
        case horizontal
        case vertical
    }
}

// MARK: - ActionFlowLayout

/// A wrapping flow layout for action buttons, matching Android's FlowRow behavior.
/// Buttons flow horizontally and wrap to the next row when they exceed available width.
private struct ActionFlowLayout: SwiftUI.Layout {
    let spacing: CGFloat
    let isRTL: Bool

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let result = arrange(proposal: proposal, subviews: subviews)
        return result.size
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        let result = arrange(proposal: proposal, subviews: subviews)
        for (index, position) in result.positions.enumerated() where index < subviews.count {
            let x: CGFloat
            if isRTL {
                x = bounds.maxX - position.x - result.sizes[index].width
            } else {
                x = bounds.minX + position.x
            }
            subviews[index].place(
                at: CGPoint(x: x, y: bounds.minY + position.y),
                proposal: ProposedViewSize(result.sizes[index])
            )
        }
    }

    private struct LayoutResult {
        var size: CGSize
        var positions: [CGPoint]
        var sizes: [CGSize]
    }

    private func arrange(proposal: ProposedViewSize, subviews: Subviews) -> LayoutResult {
        let maxWidth = proposal.width ?? .infinity
        var positions: [CGPoint] = []
        var sizes: [CGSize] = []
        var currentX: CGFloat = 0
        var currentY: CGFloat = 0
        var rowHeight: CGFloat = 0
        var totalWidth: CGFloat = 0

        // Cap each item to 1/3 of available width when 3+ items, matching Android's 3-column FlowRow
        let maxItemWidth: CGFloat? = subviews.count >= 3 && maxWidth.isFinite
            ? (maxWidth - 2 * spacing) / 3
            : nil

        for subview in subviews {
            let intrinsicSize = subview.sizeThatFits(ProposedViewSize(width: nil, height: nil))
            let cappedWidth = maxItemWidth.map { min(intrinsicSize.width, $0) } ?? intrinsicSize.width
            let size = CGSize(width: cappedWidth, height: subview.sizeThatFits(ProposedViewSize(width: cappedWidth, height: nil)).height)

            if currentX + size.width > maxWidth && currentX > 0 {
                currentX = 0
                currentY += rowHeight + spacing
                rowHeight = 0
            }

            positions.append(CGPoint(x: currentX, y: currentY))
            sizes.append(size)
            rowHeight = max(rowHeight, size.height)
            currentX += size.width + spacing
            totalWidth = max(totalWidth, currentX - spacing)
        }

        let totalHeight = currentY + rowHeight
        return LayoutResult(
            size: CGSize(width: totalWidth, height: totalHeight),
            positions: positions,
            sizes: sizes
        )
    }
}

File: android/ac-rendering/src/main/kotlin/com/microsoft/adaptivecards/rendering/composables/ActionSetView.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.rendering.composables

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material.icons.outlined.*
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.microsoft.adaptivecards.core.models.*
import com.microsoft.adaptivecards.rendering.theme.LocalHostConfig
import com.microsoft.adaptivecards.rendering.viewmodel.ActionHandler
import com.microsoft.adaptivecards.rendering.viewmodel.CardViewModel
import com.microsoft.adaptivecards.accessibility.buttonSemantics

/**
 * Renders an ActionSet as a row or column of action buttons.
 *
 * Separates actions into primary (visible buttons) and secondary (overflow menu)
 * based on the action `mode` property. Primary actions exceeding `maxActions` from
 * HostConfig are also moved into the overflow menu, matching the web renderer behavior.
 */
@Composable
fun ActionSetView(
    actions: List<CardAction>,
    actionHandler: ActionHandler,
    viewModel: CardViewModel,
    modifier: Modifier = Modifier
) {
    val hostConfig = LocalHostConfig.current
    val maxActions = hostConfig.actions.maxActions

    // Separate actions by mode: Primary (default) vs Secondary (overflow)
    val primaryActions = actions.filter { it.mode != ActionMode.Secondary }
    val secondaryActions = actions.filter { it.mode == ActionMode.Secondary }.toMutableList()

    // If primary actions exceed maxActions, move the excess into the overflow menu
    val visibleActions: List<CardAction>
    if (primaryActions.size > maxActions) {
        visibleActions = primaryActions.take(maxActions)
        secondaryActions.addAll(0, primaryActions.drop(maxActions))
    } else {
        visibleActions = primaryActions
    }

    val isLeftAligned = hostConfig.actions.actionAlignment == "left"

    Column(modifier = modifier) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(
                hostConfig.actions.buttonSpacing.dp
            )
        ) {
            visibleActions.forEach { action ->
                ActionButton(
                    action = action,
                    actionHandler = actionHandler,
                    viewModel = viewModel,
                    modifier = if (isLeftAligned) Modifier else Modifier.weight(1f)
                )
            }

            // Overflow menu for secondary actions
            if (secondaryActions.isNotEmpty()) {
                OverflowMenuButton(
                    actions = secondaryActions,
                    actionHandler = actionHandler,
                    viewModel = viewModel
                )
            }
        }

        // Render expanded ShowCard sub-cards inline
        val showCardActions = actions.filterIsInstance<ActionShowCard>()
        showCardActions.forEach { showCardAction ->
            val actionId = showCardAction.id ?: "showCard_${showCardAction.title ?: "unknown"}"
            if (viewModel.isShowCardExpanded(actionId)) {
                val emphasisBg = Color(
                    android.graphics.Color.parseColor(
                        hostConfig.containerStyles.emphasis.backgroundColor
                    )
                )
                val cornerRadius = hostConfig.cornerRadius.container.dp

                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = hostConfig.spacing.default.dp)
                        .background(
                            color = emphasisBg,
                            shape = RoundedCornerShape(cornerRadius)
                        )
                        .padding(hostConfig.spacing.padding.dp)
                ) {
                    showCardAction.card.body?.forEachIndexed { index, element ->
                        RenderElement(
                            element = element,
                            isFirst = index == 0,
                            viewModel = viewModel,
                            actionHandler = actionHandler
                        )
                    }
                    // Render sub-card actions if present
                    showCardAction.card.actions?.let { subActions ->
                        if (subActions.isNotEmpty()) {
                            Spacer(modifier = Modifier.height(hostConfig.spacing.default.dp))
                            ActionSetView(
                                actions = subActions,
                                actionHandler = actionHandler,
                                viewModel = viewModel
                            )
                        }
                    }
                }
            }
        }

        // Render popover bottom sheets for any ActionPopover actions that are showing
        val popoverActions = actions.filterIsInstance<ActionPopover>()
        popoverActions.forEach { popoverAction ->
            val actionId = popoverAction.id ?: "popover_${popoverAction.title ?: "unknown"}"
            if (viewModel.isPopoverShowing(actionId)) {
                PopoverBottomSheet(
                    action = popoverAction,
                    viewModel = viewModel,
                    actionHandler = actionHandler,
                    onDismiss = { viewModel.dismissPopover(actionId) }
                )
            }
        }
    }
}

/**
 * Overflow "..." button that opens a dropdown menu with secondary actions.
 */
@Composable
private fun OverflowMenuButton(
    actions: List<CardAction>,
    actionHandler: ActionHandler,
    viewModel: CardViewModel
) {
    var expanded by remember { mutableStateOf(false) }

    Box {
        OutlinedButton(
            onClick = { expanded = true },
            modifier = Modifier.semantics {
                contentDescription = "More actions"
            }
        ) {
            Text("\u2026") // Ellipsis character
        }

        DropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false }
        ) {
            actions.forEach { action ->
                DropdownMenuItem(
                    text = { Text(action.title ?: "") },
                    onClick = {
                        expanded = false
                        handleAction(action, actionHandler, viewModel)
                    },
                    enabled = action.isEnabled
                )
            }
        }
    }
}

/**
 * Renders a single action as a button.
 *
 * Default style: outlined with accent border (matching Figma).
 * Positive/Destructive: filled with semantic color.
 * Icons placed to the left of title per HostConfig.actions.iconPlacement.
 */
@Composable
fun ActionButton(
    action: CardAction,
    actionHandler: ActionHandler,
    viewModel: CardViewModel,
    modifier: Modifier = Modifier
) {
    val hostConfig = LocalHostConfig.current
    val accentColor = Color(
        android.graphics.Color.parseColor(
            hostConfig.containerStyles.default.foregroundColors.accent.default
        )
    )

    // Use tooltip as content description for accessibility when available
    val tooltipText = action.tooltip
    val cornerRadius = hostConfig.cornerRadius.container.dp
    val buttonShape = RoundedCornerShape(cornerRadius)
    val buttonPadding = PaddingValues(horizontal = hostConfig.spacing.medium.dp, vertical = (hostConfig.spacing.small * 0.75f).dp)

    when (action.style) {
        ActionStyle.Positive -> {
            val goodColor = Color(android.graphics.Color.parseColor(hostConfig.containerStyles.default.foregroundColors.good.default))
            Button(
                onClick = { handleAction(action, actionHandler, viewModel) },
                enabled = action.isEnabled,
                colors = ButtonDefaults.buttonColors(containerColor = goodColor),
                shape = buttonShape,
                contentPadding = buttonPadding,
                modifier = modifier.buttonSemantics(
                    label = tooltipText ?: action.title ?: "Action",
                    enabled = action.isEnabled
                )
            ) {
                ActionButtonContent(action, hostConfig)
            }
        }
        ActionStyle.Destructive -> {
            val attentionColor = Color(android.graphics.Color.parseColor(hostConfig.containerStyles.default.foregroundColors.attention.default))
            Button(
                onClick = { handleAction(action, actionHandler, viewModel) },
                enabled = action.isEnabled,
                colors = ButtonDefaults.buttonColors(containerColor = attentionColor),
                shape = buttonShape,
                contentPadding = buttonPadding,
                modifier = modifier.buttonSemantics(
                    label = tooltipText ?: action.title ?: "Action",
                    enabled = action.isEnabled
                )
            ) {
                ActionButtonContent(action, hostConfig)
            }
        }
        else -> {
            // Default style: outlined button with accent border
            OutlinedButton(
                onClick = { handleAction(action, actionHandler, viewModel) },
                enabled = action.isEnabled,
                border = BorderStroke(hostConfig.separator.lineThickness.dp, accentColor),
                shape = buttonShape,
                contentPadding = buttonPadding,
                colors = ButtonDefaults.outlinedButtonColors(
                    contentColor = accentColor
                ),
                modifier = modifier.buttonSemantics(
                    label = tooltipText ?: action.title ?: "Action",
                    enabled = action.isEnabled
                )
            ) {
                ActionButtonContent(action, hostConfig)
            }
        }
    }
}

/** Icon + title content shared by all button styles. */
@Composable
private fun ActionButtonContent(
    action: CardAction,
    hostConfig: com.microsoft.adaptivecards.core.hostconfig.HostConfig
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(hostConfig.spacing.small.dp),
        verticalAlignment = androidx.compose.ui.Alignment.CenterVertically
    ) {
        action.iconUrl?.let { iconUrl ->
            val iconSize = hostConfig.actions.iconSize.dp
            if (iconUrl.startsWith("icon:")) {
                val iconVector = resolveFluentIcon(iconUrl.removePrefix("icon:"))
                if (iconVector != null) {
                    Icon(
                        imageVector = iconVector,
                        contentDescription = null,
                        modifier = Modifier.size(iconSize)
                    )
                }
            } else {
                AsyncImage(
                    model = iconUrl,
                    contentDescription = null,
                    modifier = Modifier.size(iconSize)
                )
            }
        }
        Text(action.title ?: "")
    }
}

/** Maps Fluent UI icon names to Material icons. Strips style suffixes like ",Filled" or ",Regular". */
private fun resolveFluentIcon(name: String): ImageVector? {
    // Strip style suffix (e.g., ",Filled", ",Regular") before lookup
    val baseName = name.split(",").firstOrNull()?.lowercase() ?: return null
    return when (baseName) {
        "alerturgent" -> Icons.Filled.Notifications
        "alert", "bell" -> Icons.Outlined.Notifications
        "belloff" -> Icons.Outlined.Notifications
        "calendar" -> Icons.Outlined.DateRange
        "send" -> Icons.Filled.Send
        "edit" -> Icons.Outlined.Edit
        "delete" -> Icons.Outlined.Delete
        "add" -> Icons.Filled.Add
        "search" -> Icons.Filled.Search
        "share" -> Icons.Filled.Share
        "star" -> Icons.Outlined.Star
        "starfilled" -> Icons.Filled.Star
        "heart" -> Icons.Outlined.FavoriteBorder
        "heartfilled" -> Icons.Filled.Favorite
        "bookmark" -> Icons.Outlined.BookmarkBorder
        "bookmarkfilled" -> Icons.Filled.Bookmark
        "link" -> Icons.Filled.Share
        "clock" -> Icons.Filled.DateRange
        "comment" -> Icons.Filled.Email
        "thumblike" -> Icons.Outlined.ThumbUp
        "eye" -> Icons.Filled.Visibility
        "eyeoff" -> Icons.Filled.VisibilityOff
        "checkmarkcircle" -> Icons.Outlined.CheckCircle
        "dismisscircle" -> Icons.Filled.Clear
        "info" -> Icons.Outlined.Info
        "warning" -> Icons.Outlined.Warning
        "errorcircle" -> Icons.Filled.Error
        "open" -> Icons.Filled.OpenInNew
        "copy" -> Icons.Filled.ContentCopy
        "save" -> Icons.Filled.Done
        "flag" -> Icons.Outlined.Flag
        "flagfilled" -> Icons.Filled.Flag
        "location" -> Icons.Outlined.LocationOn
        "phone", "call" -> Icons.Filled.Call
        "mail" -> Icons.Outlined.Email
        "video" -> Icons.Filled.PlayArrow
        "camera" -> Icons.Filled.CameraAlt
        "attach" -> Icons.Filled.AttachFile
        "document" -> Icons.Filled.Description
        "folder" -> Icons.Filled.Folder
        "settings" -> Icons.Outlined.Settings
        "filter" -> Icons.Filled.FilterList
        "morehorizontal" -> Icons.Filled.MoreHoriz
        "chevronright" -> Icons.Filled.ChevronRight
        "chevrondown" -> Icons.Filled.ExpandMore
        "chevronup" -> Icons.Filled.ExpandLess
        "navigation" -> Icons.Filled.Navigation
        "receipt" -> Icons.Filled.Receipt
        "arrowreset" -> Icons.Filled.Refresh
        "toggleleft" -> Icons.Filled.ToggleOn
        else -> null
    }
}

/**
 * Renders an ActionPopover's content inside a Material3 ModalBottomSheet.
 *
 * The bottom sheet does not steal focus and supports nested popovers
 * (each popover tracks its own state via viewModel.popoverState).
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun PopoverBottomSheet(
    action: ActionPopover,
    viewModel: CardViewModel,
    actionHandler: ActionHandler,
    onDismiss: () -> Unit
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val hostConfig = LocalHostConfig.current

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = hostConfig.spacing.padding.dp)
                .padding(bottom = hostConfig.spacing.padding.dp)
        ) {
            // Popover title
            val title = action.popoverTitle ?: action.title
            if (!title.isNullOrBlank()) {
                Text(
                    text = title,
                    fontSize = 18.sp,
                    fontWeight = FontWeight.SemiBold,
                    modifier = Modifier.padding(bottom = hostConfig.spacing.default.dp)
                )
            }

            // Render the popover content element using the standard rendering pipeline
            action.content?.let { contentElement ->
                RenderElement(
                    element = contentElement,
                    isFirst = true,
                    viewModel = viewModel,
                    actionHandler = actionHandler
                )
            }
        }
    }
}

/**
 * Handle action execution
 */
private fun handleAction(
    action: CardAction,
    actionHandler: ActionHandler,
    viewModel: CardViewModel
) {
    when (action) {
        is ActionSubmit -> {
            val inputData = viewModel.getAllInputValues()
            actionHandler.onSubmit(inputData, action.id)
        }
        is ActionOpenUrl -> {
            actionHandler.onOpenUrl(action.url, action.id)
        }
        is ActionExecute -> {
            val inputData = viewModel.getAllInputValues()
            actionHandler.onExecute(action.verb ?: "", inputData, action.id)
        }
        is ActionShowCard -> {
            val actionId = action.id ?: "showCard_${action.title ?: "unknown"}"
            viewModel.toggleShowCard(actionId)
            actionHandler.onShowCard(action)
        }
        is ActionToggleVisibility -> {
            action.targetElements.forEach { target ->
                viewModel.toggleVisibility(target.elementId, target.isVisible)
            }
            actionHandler.onToggleVisibility(action.targetElements.map { it.elementId })
        }
        is ActionOpenUrlDialog -> {
            actionHandler.onOpenUrl(action.url, action.id)
        }
        is ActionPopover -> {
            val actionId = action.id ?: "popover_${action.title ?: "unknown"}"
            viewModel.togglePopover(actionId)
        }
        is ActionRunCommands -> {
            // RunCommands actions handled externally
        }
    }
}
